pages/2_subscriptions_report.py

```diff
@@ -18,46 +19,293 @@
+data['subscription_started_month'] = data['subscription_period_started_at'].dt.to_period('M')
+data['subscription_started_quarter'] = data['subscription_period_started_at'].dt.to_period('Q')
+
+# Filter the Dataframe to include only 'subscription' and 'recurring' billing types
+subscriptions_data = data[
+    (data['billing_type'].isin(['subscription', 'recurring'])) & 
+    (data['transaction_type'] == 'sale')
+]
+
+# Filter so payment_month is within the date range
+subscriptions_revenue_data = subscriptions_data[
+    (subscriptions_data['payment_month'] >= min_date) &
+    (subscriptions_data['payment_month'] <= max_date)
+]
+
+# Filter the Dataframe to include only 'one-time' and 'invoiceitem' billing types
+onetime_data = data[
+    (data['billing_type'].isin(['one-time', 'invoiceitem'])) & 
+    (data['transaction_type'] == 'sale')
+]
+
+# Active subscriptions data
+active_subscriptions_data = subscriptions_data[subscriptions_data['subscription_status'] == 'active']
+
+# Filter subscription_started_month to be within the min and max dates
+subscriptions_starts_data = subscriptions_data[
+    (subscriptions_data['subscription_started_month'] >= min_date) &
+    (subscriptions_data['subscription_started_month'] <= max_date)
+]
+
+# Filter subscriptions_starts_data to active only subscriptions
+new_subscriptions_data = subscriptions_starts_data[subscriptions_starts_data['subscription_status'] == 'active']
+
+# Group by 'payment_month' to calculate MRR and single order
+mrr_data = subscriptions_data.groupby(subscriptions_data['payment_month'])['total_amount'].sum().sort_index()
+single_order_data = onetime_data.groupby(onetime_data['payment_month'])['total_amount'].sum().sort_index()
+
+# Filter to be within the min_date and max_date
+mrr_data = mrr_data[
+    (mrr_data.index >= min_date) & 
+    (mrr_data.index <= max_date)
+]
+single_order_data = single_order_data[
+    (single_order_data.index >= min_date) & 
+    (single_order_data.index <= max_date)
+]
+
 ## KPI Metrics which need to be updated.
 with st.container():
     col1, col2, col3, col4, col5 = st.columns(5)
 
+    # Total Revenue From Subscriptions
     with col1:
-        st.metric(label="**Total Revenue From Subscriptions**", value="TBD")
+        current_total_revenue = subscriptions_data['total_amount'].sum()
+        last_quarter_total_revenue = subscriptions_data[subscriptions_data['payment_month'].dt.quarter == (subscriptions_data['payment_month'].dt.quarter.max() - 1)]['total_amount'].sum()
+        current_total_revenue_str = f'${current_total_revenue:,.2f}' if not pd.isna(current_total_revenue) else "no data"
+        qoq_total_revenue = current_total_revenue - last_quarter_total_revenue
+        qoq_total_revenue_str = f'{qoq_total_revenue:,.2f} QoQ' if not pd.isna(qoq_total_revenue) else "no data"
+        st.metric(
+            label="**Total Revenue From Subscriptions**", 
+            value=current_total_revenue_str,
+            delta=qoq_total_revenue_str
```

Comment:
I apologize for this flip flop request, but we found for the Churn and Order/Revenue that QoQ with our fake data was telling a confusing story for the demo purposes. I know I requested originally that you change these to QoQ, but given the updates made to the other pages would you be able to switch this back to YoY to remain consistent with the other pages.

Reply:
Updated.